core/chains/evm/logpoller/expression.go

```diff
@@ -0,0 +1,147 @@
+package logpoller
+
+import (
+	"fmt"
+	"math/big"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query"
+)
+
+type qSelect struct {
+	Fields []string
+	Table  string
+	Limit  query.Limit
+}
+
+func (q qSelect) String() string {
+	fields := "*"
+	fieldsOut := q.Fields
+
+	if hasCursorLimit(q.Limit) {
+		fieldsOut = append(fieldsOut, fmt.Sprintf("%s AS %s", cursorFieldName, cursorFieldAlias))
+	}
+
+	if len(fieldsOut) > 0 {
+		fields = strings.Join(fieldsOut, ",")
+	}
+
+	return fmt.Sprintf("SELECT %s FROM %s", fields, q.Table)
+}
+
+type qWhere struct {
+	ChainID     *big.Int
+	Address     common.Address
+	Op          query.BoolOperator
+	Expressions []query.Expression
+	Limit       query.Limit
+}
+
+func (q qWhere) String() (string, map[string]any) {
+	args := newNamedArgs()
+
+	segment := fmt.Sprintf(
+		"WHERE evm_chain_id = %s AND address = %s AND %s",
+		args.Add("evm_chain_id", q.ChainID.Uint64()),
+		args.Add("address", q.Address),
+		makeExpression(q.Op, q.Expressions, args, q.ChainID),
+	)
+
+	if hasCursorLimit(q.Limit) {
+		var op string
+		switch *q.Limit.CursorDirection {
+		case query.Following:
+			op = ">"
+		case query.Previous:
+			op = "<"
+		default:
+			op = "<"
+		}
+
+		segment = fmt.Sprintf(
+			"%s AND %s %s %s",
+			segment,
+			cursorFieldAlias,
+			op,
+			args.Add(cursorFieldAlias, *q.Limit.Cursor),
+		)
+	}
+
+	return segment, args.Values()
+}
+
+type qOrder []query.SortBy
+
+func (q qOrder) String() string {
+	if len(q) == 0 {
+		return ""
+	}
+
+	sort := make([]string, len(q))
+
+	for idx, sorted := range q {
+		var name string
+
+		switch sorted.(type) {
+		case query.SortByBlock:
+			name = blockFieldName
+		case query.SortBySequence:
```

Comment:
This should translate to whatever uniquely identifies the order of things on that chain. Something like block-txID-logID... whatever the EVM equivalent is

Reply:
Cursor is a separate thing that should be that plus a proper psql cursor? Not completely sure how that looks in practice, but the use case is product can tell the chain reader continue from here